crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,248 @@
+use crate::price_estimation::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use anyhow::Result;
+use futures::future::Shared;
+use futures::FutureExt;
+use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+
+type EstimationResult = Result<Estimate, PriceEstimationError>;
+type SharedEstimationRequest = Shared<Pin<Box<dyn Future<Output = EstimationResult> + Send>>>;
+
+struct Inner {
+    estimator: Box<dyn PriceEstimating>,
+    in_flight_requests: Mutex<HashMap<Query, SharedEstimationRequest>>,
+}
+
+impl Inner {
+    fn collect_garbage(&self) {
+        // TODO: Refactor this to use `HashMap::drain_filter` when it's stable:
+        // https://github.com/rust-lang/rust/issues/59618
+        let mut active_requests = self.in_flight_requests.lock().unwrap();
+        let completed_and_ignored_requests: Vec<_> = active_requests
+            .iter()
+            .filter_map(|(query, handle)| {
+                if matches!(handle.strong_count(), Some(1)) {
```

Comment:
Should this maybe also include:
```suggestion
                if handle.strong_count().unwrap_or_default() <= 1 {
```

I don't think it can happen in practice though... Maybe I'm just paranoid.

Reply:
This would be a very slight change in behavior if I'm reading the [docs](https://docs.rs/futures/latest/futures/future/struct.Shared.html#method.strong_count) right.
If future changes cause the `BufferingPriceEstimator` to poll the stored futures to completion, `strong_count()` would return `None` which would cause the future to get dropped regardless of other futures depending on it.
I'm not sure if that would cause problems, but probably not. 🤔

Reply:
I think this cannot happen because we never poll this future directly, we just give out copies. If that is true then I would say 

```rust
 if handle.strong_count().unwrap() == 1 {
```

is the most appropriate because it asserts this invariant.

Reply:
Based on your comment maybe I misunderstood how this works.

Reply:
I think you understood it right. In my comment I was worried about a possible change in the future.
I agree with you that right now `unwrap()` seems like the best option.

Reply:
If I'm not mistaken switching to `WeakShared` would not change the behavior of the `BufferingPriceEstimator` while better communicating the intent.

Reply:
I read the docs as `None` means that your particular instance of `Shared` has already been polled (not that the actual shared future has resolved, i.e. `shared.peek().is_some())`).

I tried it out and what I found seems to confirm this:
```rust
use futures::prelude::*;
use std::{
    pin::Pin,
    task::{Context, Poll},
};

fn main() {
    let a = async { 42_i64 }.shared();
    let mut b = a.clone();

    dbg!(a.strong_count()); // Some(2)
    dbg!(b.strong_count()); // Some(2)

    let waker = futures::task::noop_waker();
    let mut context = Context::from_waker(&waker);
    assert_eq!(Pin::new(&mut b).poll(&mut context), Poll::Ready(42));

    dbg!(a.peek()); // Some(42)
    dbg!(a.strong_count()); // Some(1)
    dbg!(b.strong_count()); // None
}
```

Reply:
I agree with this interpretation and under it I think my comment is correct that we can do unwrap. The WeakShared would communicate the intent better because it doesn't implement Future. Using it we would need to call upgrade when finding existing futures and react to None.

Reply:
One nicety of using `WeakShared` is that if we get `None` when we try to upgrade, this would mean that the future stopped being polled/resolved for all `Shared` references. In this case it would probably be good to re-query.

Reply:
One downside of `WeakShared` is that it might look a little awkward.
We would have to keep the `Shared` futures around to the end of `buffered_estimate()` only to prevent the ref counted future from being dropped immediately. Besides that we wouldn't do anything with them.
I guess I can make a commit switching to `WeakShared` and we can decide whether to keep it or revert it.

Reply:
Ah I see, because we await `fetch_remaining_estimates` and not the shared futures that get added to the `in_flight_requests.`